fscache.go

```diff
@@ -0,0 +1,698 @@
+	ops_garbage_collected uint64
+	// Number of in-progress alloc_object() cache ops
+	cacheop_allocations_in_progress uint64
+	// Number of in-progress lookup_object() cache ops
+	cacheop_lookup_object_in_progress uint64
+	// Number of in-progress lookup_complete() cache ops
+	cacheop_lookup_complete_in_progress uint64
+	// Number of in-progress grab_object() cache ops
+	cacheop_grab_object_in_progress uint64
+	cacheop_invalidations           uint64
+	// Number of in-progress update_object() cache ops
+	cacheop_update_object_in_progress uint64
+	// Number of in-progress drop_object() cache ops
+	cacheop_drop_object_in_progress uint64
+	// Number of in-progress put_object() cache ops
+	cacheop_put_object_in_progress uint64
+	// Number of in-progress attr_changed() cache ops
+	cacheop_attribute_change_in_progress uint64
+	// Number of in-progress sync_cache() cache ops
+	cacheop_sync_cache_in_progress uint64
+	// Number of in-progress read_or_alloc_page() cache ops
+	cacheop_read_or_alloc_page_in_progress uint64
+	// Number of in-progress read_or_alloc_pages() cache ops
+	cacheop_read_or_alloc_pages_in_progress uint64
+	// Number of in-progress allocate_page() cache ops
+	cacheop_allocate_page_in_progress uint64
+	// Number of in-progress allocate_pages() cache ops
+	cacheop_allocate_pages_in_progress uint64
+	// Number of in-progress write_page() cache ops
+	cacheop_write_pages_in_progress uint64
+	// Number of in-progress uncache_page() cache ops
+	cacheop_uncache_pages_in_progress uint64
+	// Number of in-progress dissociate_pages() cache ops
+	cacheop_dissociate_pages_in_progress uint64
+	// Number of object lookups/creations rejected due to lack of space
+	cacheev_lookups_and_creations_rejected_lack_space uint64
+	// Number of stale objects deleted
+	cacheev_stale_objects_deleted uint64
+	// Number of objects retired when relinquished
+	cacheev_retired_when_relinquished uint64
+	// Number of objects culled
+	cacheev_objects_culled uint64
+}
+
+// Meminfo returns an information about current kernel/system memory statistics.
+// See https://www.kernel.org/doc/Documentation/filesystems/proc.txt
+func (fs FS) Fscacheinfo() (Fscacheinfo, error) {
+	b, err := util.ReadFileNoStat(fs.proc.Path("fs/fscache/stats"))
+	if err != nil {
+		return Fscacheinfo{}, err
+	}
+
+	m, err := parseFscacheinfo(bytes.NewReader(b))
+	if err != nil {
+		return Fscacheinfo{}, fmt.Errorf("failed to parse Fscacheinfo: %v", err)
+	}
+
+	return *m, nil
+}
+
+func parseFscacheinfo(r io.Reader) (*Fscacheinfo, error) {
+	var m Fscacheinfo
+	var err error
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		// Each line has at least a name and value; we ignore the unit.
+		fields := strings.Fields(s.Text())
+		if len(fields) < 2 {
+			return nil, fmt.Errorf("malformed Fscacheinfo line: %q", s.Text())
+		}
+
+		switch fields[0] {
+		case "Cookies:":
+			m.index_cookies_allocated, err = strconv.ParseUint(strings.Split(fields[1], "=")[1], 0, 64)
```

Comment:
There appears to be some risk of running into array index out of bounds exception here if there are fewer fields than expected.  I'm wondering if these cases could be refactored using a function that checks for the expected number of fields and then sets them.

```golang
func setFSCacheFields(fields []string, setFields  ...*uint64) error {
    if len(fields) < len(setFields) {
        return fmt.Errorf("Insufficient number of fields, expected %v, got %v", len(setFields), len(fields))
    }
    for i := range setFields {
        *setFields[i], err = strconv.ParseUint(strings.Split(fields[i], "=")[1], 0, 64)
        if err != nil {
            return error
        }
    }
}
```

Then for each case, you just pass the fields array and the list of fields to be set.
```golang
    err := setFSCacheFields(fields, &m.index_cookies_allocated, &m.data_storage_cookies_allocated, &m.special_cookies_allocated)
```

WDYT?

Reply:
Sure, I've changed to your suggested method, which is much safer.